scripts/rescript_format.js

```diff
@@ -118,7 +118,10 @@ function main(argv, rescript_exe, bsc_exe) {
         );
         (async function () {
           var content = await readStdin();
-          fs.writeFileSync(filename, content, "utf8");
+          var fd = fs.openSync(filename, 'wx', 0o600);
+          fs.writeFileSync(fd, content, "utf8");
```

Comment:
does this happen in practice? note the filename already used some crypto randomness

Reply:
I think the chance of choosing a file that already exists is very low, but the old code has a bug/security issue called [TOCTOU](https://en.m.wikipedia.org/wiki/Time-of-check_to_time-of-use). It is better not to risk it.

Reply:
Hi, we want spend the limited time in solving the most pressing problems. This commit makes it safer in theory but not that much in practice. And there is a cost for it, e.g, you and me are reviewing the code on weekend time.
If you can come with a realworld issue, I am happy to revisit it.

Reply:
[CWE-378: Creation of Temporary File With Insecure Permissions](https://cwe.mitre.org/data/definitions/378.html) goes into the some of the issues.

Reply:
The random bytes + umask 022 make it infeasible (but not impossible) to do injection. However, this still leaks the input to other (non-root) users on the system, which isn't great (especially on shared/CI systems running potentially untrusted code).

The practical attack for that is as follows:
1. subscribe to file write/creation in /tmp using inotify
2. read and save /tmp/rescript_* files.

If the `unlinkSync` call is new, you can actually do this long after process exit.

Reply:
@Minnozz would you mind put the controversial diff in a separate diff? other parts look good to me

Reply:
@bobzhang Done: #5220